google-cloud-pubsub/lib/google/cloud/pubsub/publisher.rb

```diff
@@ -0,0 +1,195 @@
+require "google/cloud/pubsub/service"
+
+module Google
+  module Cloud
+    module Pubsub
+      ##
+      # # Publisher
+      #
+      class Publisher
+        include MonitorMixin
+
+        MAX_COUNT = 2000
+        MAX_SIZE = 10000000
+
+        attr_reader :service, :queue
+        attr_reader :max_size, :max_count
+        attr_reader :thread_pool, :publish_task
+
+        def initialize service, max_size: MAX_SIZE, max_count: MAX_COUNT
+          @service = service
+          @max_size = max_size
+          @max_count = max_count
+
+          @queue = {}
+          @thread_pool = Concurrent::FixedThreadPool.new(
+            [2, Concurrent.processor_count * 2].max,
+            fallback_policy: :caller_runs
+          )
+          @publish_task = Concurrent::TimerTask.new(execution_interval: 0.25) do
+            flush
+          end
+
+          # init MonitorMixin
+          super()
+        end
+
+        def publish topic_name, data = nil, attributes = {}, &block
+          topic_name = service.topic_path topic_name
+          msg = create_pubsub_message data, attributes
+
+          synchronize do
+            batch = @queue[topic_name]
+            if batch.nil?
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            elsif !batch.try_add(msg, block, max_count: max_count,
+                                             max_size: max_size)
+              # We hit a threshold, publish existing batch.
+              publish_batch_async topic_name, batch
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            end
+            @queue[topic_name] = batch
+          end
+          nil
+        end
+
+        def start
+          @publish_task.execute
+        end
+
+        def stop
+          flush
+          @publish_task.shutdown
+        end
+
+        def started?
+          @publish_task.running?
+        end
+
+        def stopping?
+          @publish_task.shuttingdown?
+        end
+
+        def stopped?
+          @publish_task.shutdown?
+        end
+
+        def wait timeout = nil
+          @publish_task.wait_for_termination timeout
+        end
+
+        def flush
+          synchronize do
+            @queue.each do |topic_name, batch|
+              publish_batch_async topic_name, batch
+            end
+            @queue = {}
+          end
+        end
+
+        protected
+
+        def create_pubsub_message data, attributes
+          attributes ||= {}
+          if data.is_a?(::Hash) && attributes.empty?
+            attributes = data
+            data = nil
+          end
+
+          # Convert IO-ish objects to strings
+          if data.respond_to?(:read) && data.respond_to?(:rewind)
+            data.rewind
+            data = data.read
+          end
+          # Convert data to encoded byte array to match the protobuf defn
+          data = String(data).force_encoding("ASCII-8BIT")
```

Comment:
Summoning @remi. :-) So, disclaimer: I am a Python guy so I have Python biases.

That said, I think this should be an error rather than an implicit conversion. Same for line 128 below.
This is a hard spot, though, because I know that being implicit and Postel's Law are important in Ruby.

Here is why I think this might be important enough to break Ruby convention: Bear in mind that with Pub/Sub, you do not know anything about the client on the other side. They may not (in fact, likely will not) be using this library, or Ruby at all. You can not assume _anything_ about the folks on the consuming side of this. Well, that is not _quite_ true: You can basically assume that the client will base64 decode the message, and nothing else.

I am doing a quick bit of research on string encoding in Ruby (since I do not know much about the ins and outs of this topic in Ruby; note that I _am_ fluent on the topic more generally -- an entire chapter of my book was devoted to this issue), and as best as I can tell, Ruby [basically has the same problems that everyone else has][1]. Basically, strings have an explicit `encoding`, but it might be wrong, especially if you read in from a file.

This merits discussion. But I _think_ that users need to send the right thing, and we should actually complain loudly rather than coerce for them, because the fundamental awareness here is so critical.

  [1]: http://www.justinweiss.com/articles/3-steps-to-fix-encoding-problems-in-ruby/